src/main/scala/io/iohk/ethereum/consensus/ConsensusBuilder.scala

```diff
@@ -1,58 +1,65 @@
 package io.iohk.ethereum.consensus
 
 import io.iohk.ethereum.consensus.atomixraft.AtomixRaftConsensus
-import io.iohk.ethereum.nodebuilder.{BlockchainConfigBuilder, ShutdownHookBuilder}
+import io.iohk.ethereum.consensus.ethash.EthashConsensus
+import io.iohk.ethereum.nodebuilder._
 import io.iohk.ethereum.utils.{Config, Logger}
 
 /**
  * A consensus builder is responsible to instantiate the consensus protocol.
  * This is done dynamically when Mantis boots, based on its configuration.
+ *
+ * @see [[io.iohk.ethereum.consensus.Consensus Consensus]],
+ *      [[io.iohk.ethereum.consensus.ethash.EthashConsensus EthashConsensus]],
+ *      [[io.iohk.ethereum.consensus.atomixraft.AtomixRaftConsensus AtomixRaftConsensus]]
  */
 trait ConsensusBuilder {
-  self: BlockchainConfigBuilder with ConsensusConfigBuilder with Logger =>
+  self: VmBuilder with BlockchainBuilder with BlockchainConfigBuilder with ConsensusConfigBuilder with Logger ⇒
 
   private lazy val mantisConfig = Config.config
 
-  private def loadEthashConsensus(): ethash.EthashConsensus = {
-    val miningConfig = ethash.MiningConfig(mantisConfig)
-    val fullConfig = FullConsensusConfig(consensusConfig, miningConfig)
-    val consensus = new ethash.EthashConsensus(blockchainConfig, fullConfig)
-    consensus
-  }
+  private def newConfig[C <: AnyRef](c: C): FullConsensusConfig[C] =
+    FullConsensusConfig(consensusConfig, c)
 
-  private def loadDemoConsensus(): demo.DemoConsensus = {
-    val demoConsensusConfig = demo.DemoConsensusConfig(mantisConfig)
-    val fullConfig = FullConsensusConfig(consensusConfig, demoConsensusConfig)
-    val consensus = new demo.DemoConsensus(blockchainConfig, fullConfig)
+  protected def buildEthashConsensus(): ethash.EthashConsensus = {
+    val specificConfig = ethash.EthashConfig(mantisConfig)
+    val fullConfig = newConfig(specificConfig)
+    val consensus = EthashConsensus(vm, blockchain, blockchainConfig, fullConfig)
     consensus
   }
 
-  private def loadAtomixRaftConsensus(): atomixraft.AtomixRaftConsensus = {
-    val atomixRaftConfig = atomixraft.AtomixRaftConfig(mantisConfig)
-    val fullConfig = FullConsensusConfig(consensusConfig, atomixRaftConfig)
-    val consensus = new AtomixRaftConsensus(blockchainConfig, fullConfig)
+  protected def buildAtomixRaftConsensus(): atomixraft.AtomixRaftConsensus = {
+    val specificConfig = atomixraft.AtomixRaftConfig(mantisConfig)
+    val fullConfig = newConfig(specificConfig)
+    val consensus = AtomixRaftConsensus(vm, blockchain, blockchainConfig, fullConfig)
     consensus
   }
 
-  private def loadConsensus(): Consensus = {
+  protected def buildConsensus(): TestConsensus = {
```

Comment:
I have mixed feelings about `TestConsensus` leaking to main app interfaces. What do you think about using `Consensus` here and casting it to `TestConsensus` in tests?

Reply:
Leaking is not good and this is a compromise trying to a void the cast. But I think there is a better way: provide a `TestConsensusBuilder` that has the correct subtype and which is only used in the tests. I'll try it and if it does not complicate the codebase, I'll commit a patch. Does that sound OK?

Reply:
OK, [implemented](https://github.com/input-output-hk/mantis/pull/410/commits/adcdccb373deb14533176fcd5cde327cdc112aea).